lib/Frontend/Frontend.cpp

```diff
@@ -440,141 +549,150 @@ void CompilerInstance::performSema() {
-    bool mainIsPrimary =
-      (PrimaryBufferID == NO_SUCH_BUFFER || MainBufferID == PrimaryBufferID);
-
-    SourceFile &MainFile =
-      MainModule->getMainSourceFile(Invocation.getSourceFileKind());
-
-    auto &Diags = MainFile.getASTContext().Diags;
-    auto DidSuppressWarnings = Diags.getSuppressWarnings();
-    Diags.setSuppressWarnings(DidSuppressWarnings || !mainIsPrimary);
-
-    SILParserState SILContext(TheSILModule.get());
-    unsigned CurTUElem = 0;
-    bool Done;
-    do {
-      // Pump the parser multiple times if necessary.  It will return early
-      // after parsing any top level code in a main module, or in SIL mode when
-      // there are chunks of swift decls (e.g. imports and types) interspersed
-      // with 'sil' definitions.
-      parseIntoSourceFile(MainFile, MainFile.getBufferID().getValue(), &Done,
-                          TheSILModule ? &SILContext : nullptr,
-                          &PersistentState, DelayedCB.get());
-      if (mainIsPrimary) {
-        performTypeChecking(MainFile, PersistentState.getTopLevelContext(),
-                            TypeCheckOptions, CurTUElem,
-                            options.WarnLongFunctionBodies,
-                            options.WarnLongExpressionTypeChecking,
-                            options.SolverExpressionTimeThreshold);
-      }
-      CurTUElem = MainFile.Decls.size();
-    } while (!Done);
-
-    Diags.setSuppressWarnings(DidSuppressWarnings);
-    
-    if (mainIsPrimary && !Context->hadError() &&
-        Invocation.getFrontendOptions().PCMacro) {
-      performPCMacro(MainFile, PersistentState.getTopLevelContext());
-    }
-    
-    // Playground transform knows to look out for PCMacro's changes and not
-    // to playground log them.
-    if (mainIsPrimary && !Context->hadError() &&
-        Invocation.getFrontendOptions().PlaygroundTransform)
-      performPlaygroundTransform(MainFile, Invocation.getFrontendOptions().PlaygroundHighPerformance);
-    if (!mainIsPrimary) {
-      performNameBinding(MainFile);
-    }
+    parseAndTypeCheckTheMainFile(PersistentState, DelayedParseCB,
+                                 TypeCheckOptions);
   }
 
-  // Type-check each top-level input besides the main source file.
-  for (auto File : MainModule->getFiles())
-    if (auto SF = dyn_cast<SourceFile>(File))
-      if (PrimaryBufferID == NO_SUCH_BUFFER || SF == PrimarySourceFile)
-        performTypeChecking(*SF, PersistentState.getTopLevelContext(),
-                            TypeCheckOptions, /*curElem*/ 0,
-                            options.WarnLongFunctionBodies,
-                            options.WarnLongExpressionTypeChecking,
-                            options.SolverExpressionTimeThreshold);
+  if (PrimaryBufferID == NO_SUCH_BUFFER)
+    typeCheckEveryFileInMainModule(PersistentState, TypeCheckOptions);
+  else if (PrimarySourceFile != nullptr)
+    typeCheckThePrimaryFile(PersistentState, TypeCheckOptions);
```

Comment:
What's in the last `else` case here? If it can't happen, we should assert instead.

Reply:
Agree! Removed the check. There is already an assert.